Adding Client Credentials to Auth

SoldierSacha opened the pull request:
<!-- Provide a brief summary of your changes -->

## Motivation and Context
<!-- Why is this change needed? What problem does it solve? -->
https://github.com/modelcontextprotocol/python-sdk/issues/881

## How Has This Been Tested?
<!-- Have you tested this in a real application? Which scenarios were tested? -->
Added test cases, and also currently using in my own mcp server

## Breaking Changes
<!-- Will users need to update their code or configurations? -->
None

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x ] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->
No

SoldierSacha commented:
@simonw @sheffler @comfuture @shimizukawa @ihrpr

SoldierSacha commented:
@Kludex @pcarleton

SoldierSacha commented:
Went ahead and updated the PR to also support the Token Exchange grant type (in addition to Client Credentials).

This is ready to be merged!! This PR is a huge win for this community. 

@dsp-ant @jspahrsummers @calclavia @nick-merrill @jerome3o-anthropic

patrykks commented:
I have some questions:

1. First of all, thanks a lot for the contribution and for preparing the MR. From my perspective, this is one of the most important features of MCP Python SDK. It's actually a key reason why a significant number (if not the majority) of companies are still not using MCP. Exposing business data without proper authentication/authorization is simply not an option in most cases.
2. Is there an ETA for when the review will be done and the code can be merged? This would be critical for our project - we’re eagerly waiting for these changes to be available
3. Is there a simple example in submitted changes showing how to configure the client credentials flow on the server side (e.g., using popular SSO providers like Azure or Okta)?

SoldierSacha commented:
> I have some questions:
> 
> 1. First of all, thanks a lot for the contribution and for preparing the MR. From my perspective, this is one of the most important features of MCP Python SDK. It's actually a key reason why a significant number (if not the majority) of companies are still not using MCP. Exposing business data without proper authentication/authorization is simply not an option in most cases.
> 2. Is there an ETA for when the review will be done and the code can be merged? This would be critical for our project - we’re eagerly waiting for these changes to be available
> 3. Is there a simple example in submitted changes showing how to configure the client credentials flow on the server side (e.g., using popular SSO providers like Azure or Okta)?

1. No problem! I needed these features myself too, so I decided to add them.

2. If you want to use these features in the meantime while the admins merge this PR, I've made a fork: https://github.com/sacha-development-stuff/mcp-python-sdk/releases/tag/v1.9.4%2Bfork

3. Regarding examples, everything can be found in /examples.

ChrisBugliosi commented:
@Kludex  @ihrpr Any news on when this will be merged? It looks like it solves some big issues with the current sdk.

patrykks commented:
Hello,

I wanted to kindly ask if there has been any progress on this issue/pull request.

I’m noticing the author has had to rebase multiple times, yet the review has still not started. In my view, the change introduced in this PR is the most important and fundamental update in the library. Without it, the SDK is practically unusable for the majority of business use cases (such as AI chatbots). I’m struggling to understand why it appears to be receiving no/low priority. Could someone from the library’s maintainers please take a look? Thank you so much.

SoldierSacha commented:
> Since #1020 was closed, I took a look at the code here, leaving a few comments:
> 
> 1. We should avoid duplicating dynamic client registration code and metadata discovery.
> 2. Client credentials has `client_secret_post` and `client_secret_basic` variants which should both be handled (this PR currently assumes `client_secret_post` without checking).
> 3. The `authorization_code`, `client_credentials`, and `token_exchange` flows should probably all extend a new base class, instead of making `ClientCredentialsProvider` extend `httpx.Auth` and making `TokenExchangeProvider` extend `ClientCredentialsProvider`.

Thank you for your comments! In my recent commits, here's what I changed to address your concerns.

1. I went ahead and implemented a shared BaseOAuthProvider that centralizes OAuth server discovery and dynamic client registration logic. This BaseOAuthProvider provides reusable helpers for all grant types.

2. I added _apply_client_auth to support both client_secret_post and client_secret_basic when authenticating to the token endpoint.

3. The token exchange is now refactored to derive from the new base class (BaseOAuthProvider) and default the resource parameter using the server URL.

felixweinberger reviewed (changes requested):
Hi @SoldierSacha thanks for this contribution! And apologies for the time it took to get back to this.

I checked with @pcarleton and it looks like this change is still pending SEP-1046: https://github.com/modelcontextprotocol/modelcontextprotocol/pull/1047

I'm going to request changes for now as there will still be merge conflict resolution and potentially minor changes needed once that SEP is accepted.

SoldierSacha commented:
Hi @felixweinberger!

I see thathttps://github.com/modelcontextprotocol/modelcontextprotocol/pull/1047 was closed in favor of https://github.com/modelcontextprotocol/ext-auth/pull/3. Would you like me to now change this as 'Ready for Review?'
